File: scripts/generate_sql_seeds.py
import os
import json
import glob
import re

def escape_sql(text):
    if not text:
        return ""
    return text.replace("'", "''")

def generate_hymns():
    print("Generating 0048_seed_hymns_from_markdown.sql...")
    output_file = "cloudflare/migrations/0048_seed_hymns_from_markdown.sql"
    
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- Migration 0048: Seed Hymns from Markdown\n")
        f.write("INSERT INTO formations (id, title, formation_type, primary_virtue, biblical_faculty, description, liturgical_script, context_anchor, min_age_months, max_age_months) VALUES\n")
        
        files = glob.glob("public/books/reformed-hymns/*.md")
        values_list = []
        
        for filepath in files:
            filename = os.path.basename(filepath)
            slug = filename.replace(".md", "").lower().replace(" ", "_")
            
            with open(filepath, "r", encoding="utf-8") as content_file:
                content = content_file.read()
                
            # Extract title (first line # Title)
            lines = content.split('\n')
            title = lines[0].replace("#", "").strip()
            if not title:
                title = filename.replace(".md", "").replace("-", " ").title()
            
            # Script is the whole content
            script = escape_sql(content)
            description = escape_sql(content[:200] + "...")
            
            values = f"('hymn_{slug}', '{escape_sql(title)}', 'liturgy', 'Worship', 'Affection', '{description}', '{script}', 'Morning_Circle', 0, 216)"
            values_list.append(values)
            
        f.write(",\n".join(values_list) + ";\n")
    print(f"Generated {len(values_list)} hymns.")

def generate_history():
    print("Generating 0049_seed_history_formations.sql...")
    output_file = "cloudflare/migrations/0049_seed_history_formations.sql"
    
    with open(output_file, "w", encoding="utf-8") as f:
        f.write("-- Migration 0049: Seed History Short Stories\n")
        f.write("INSERT INTO formations (id, title, formation_type, primary_virtue, biblical_faculty, description, liturgical_script, context_anchor, min_age_months, max_age_months) VALUES\n")
        
        dirs = glob.glob("public/books/young_historians_africa/*")
        values_list = []
        
        for dirpath in dirs:
            if not os.path.isdir(dirpath):
                continue
                
            slug = os.path.basename(dirpath)
            meta_path = os.path.join(dirpath, "metadata.json")
            content_path = os.path.join(dirpath, "content.md")
            
            if not os.path.exists(meta_path) or not os.path.exists(content_path):
                continue
                
            with open(meta_path, "r", encoding="utf-8") as meta_file:
                meta = json.load(meta_file)
                
            with open(content_path, "r", encoding="utf-8") as content_file:
                content = content_file.read()
                
            title = meta.get("title", slug.replace("_", " ").title())
            # Use content as description for the formation
            description = escape_sql(content)
            
            # Context anchor: Bedside (story time)
            values = f"('hist_story_{slug}', '{escape_sql(title)}', 'narrative', 'Wonder', 'Imagination', '{description}', '', 'Bedside', 48, 120)"
            values_list.append(values)
            
        if values_list:
            f.write(",\n".join(values_list) + ";\n")
    print(f"Generated {len(values_list)} history stories.")

if __name__ == "__main__":
    generate_hymns()
    generate_history()

File: scripts/generate_sql_seeds.js
const fs = require('fs');
const path = require('path');
const glob = require('glob'); // Assuming glob is available/installable, or use fs.readdir recursive
// Actually, standard fs.readdirSync recursively is safer if glob isn't installed.
// Let's implement a simple recursive walker to avoid dependency issues.

function getFiles(dir, files = []) {
    try {
        const fileList = fs.readdirSync(dir);
        for (const file of fileList) {
            const name = `${dir}/${file}`;
            if (fs.statSync(name).isDirectory()) {
                getFiles(name, files);
            } else {
                files.push(name);
            }
        }
    } catch (e) {
        console.warn(`Warning: Could not read directory ${dir}`);
    }
    return files;
}

function escapeSql(text) {
    if (!text) return "";
    return text.replace(/'/g, "''").replace(/\\/g, "\\\\");
}

function generateHymns() {
    console.log("Generating 0048_seed_hymns_from_markdown.sql...");
    const outputFile = "cloudflare/migrations/0048_seed_hymns_from_markdown.sql";
    const sourceDir = "public/books/reformed-hymns";

    if (!fs.existsSync(sourceDir)) {
        console.error(`Source directory not found: ${sourceDir}`);
        return;
    }

    const files = getFiles(sourceDir).filter(f => f.endsWith('.md'));
    let sql = "-- Migration 0048: Seed Hymns from Markdown\n";
    sql += "INSERT INTO formations (id, title, formation_type, primary_virtue, biblical_faculty, description, liturgical_script, context_anchor, min_age_months, max_age_months) VALUES\n";

    const values = [];

    for (const filepath of files) {
        const filename = path.basename(filepath);
        const slug = filename.replace(".md", "").toLowerCase().replace(/ /g, "_").replace(/-/g, "_");
        const content = fs.readFileSync(filepath, 'utf-8');

        const lines = content.split('\n');
        let title = lines[0].replace("#", "").trim();
        if (!title) title = filename.replace(".md", "").replace(/-/g, " ");

        const description = escapeSql(content.substring(0, 200) + "...");
        const script = escapeSql(content);

        values.push(`('hymn_${slug}', '${escapeSql(title)}', 'liturgy', 'Worship', 'Affection', '${description}', '${script}', 'Morning_Circle', 48, 216)`);
    }

    if (values.length > 0) {
        sql += values.join(",\n") + ";\n";
        fs.writeFileSync(outputFile, sql);
        console.log(`Generated ${values.length} hymns.`);
    } else {
        console.log("No hymns found.");
    }
}

function generateHistory() {
    console.log("Generating 0049_seed_history_formations.sql...");
    const outputFile = "cloudflare/migrations/0049_seed_history_formations.sql";
    const sourceDir = "public/books/young_historians_africa";

    if (!fs.existsSync(sourceDir)) {
        console.error(`Source directory not found: ${sourceDir}`);
        return;
    }

    // Get direct subdirectories
    const dirs = fs.readdirSync(sourceDir).map(d => path.join(sourceDir, d)).filter(d => fs.statSync(d).isDirectory());

    let sql = "-- Migration 0049: Seed History Short Stories\n";
    sql += "INSERT INTO formations (id, title, formation_type, primary_virtue, biblical_faculty, description, liturgical_script, context_anchor, min_age_months, max_age_months) VALUES\n";

    const values = [];

    for (const dirpath of dirs) {
        const slug = path.basename(dirpath);
        const metaPath = path.join(dirpath, "metadata.json");
        const contentPath = path.join(dirpath, "content.md");

        if (!fs.existsSync(metaPath) || !fs.existsSync(contentPath)) continue;

        const meta = JSON.parse(fs.readFileSync(metaPath, 'utf-8'));
        const content = fs.readFileSync(contentPath, 'utf-8');

        const title = meta.title || slug.replace(/_/g, " ");
        const description = escapeSql(content); // Store full content in description for now

        values.push(`('hist_story_${slug}', '${escapeSql(title)}', 'narrative', 'Wonder', 'Imagination', '${description}', '', 'Bedside', 48, 120)`);
    }

    if (values.length > 0) {
        sql += values.join(",\n") + ";\n";
        fs.writeFileSync(outputFile, sql);
        console.log(`Generated ${values.length} history stories.`);
    } else {
        console.log("No history stories found.");
    }
}

generateHymns();
generateHistory();
